### PR TITLE
Loki logging integration

### PR DESCRIPTION
**Pull request by danielwhatmuff**

### CODE REVIEW

File: app/agents/loki_agent.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+require 'cgi'
```
Comment (jits) [inline app/agents/loki_agent.rb]: Don't need this (I think).

File: app/services/resource_types_service.rb [hunk 2]
```diff
@@ -40,9 +40,17 @@ def all
           name: 'Monitoring Dashboards',
           top_level: false,
           providers: %w[grafana].freeze
+        },
+        {
+          id: 'LoggingDashboard',
+          class: 'Resources::LoggingDashboard',
+          name: 'Logging View',
```
Comment (jits) [inline app/services/resource_types_service.rb]:
```suggestion
          name: 'Logging Dashboard',
```

File: app/agents/loki_agent.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+class LokiAgent
+  def initialize(grafana_url:)
+    @grafana_url = grafana_url
+  end
+
+  def create_logging_view(namespace)
```
Comment (jits) [inline app/agents/loki_agent.rb]:
```suggestion
  def create_logging_dashboard(name)
```

File: app/agents/loki_agent.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    logging_view_uri = 'explore?left=[
+                          "now-6h",
+                          "now",
+                          "Loki",
+                          {
+                            "expr":"{namespace=\"' + namespace + '\"}
```
Comment (jits) [inline app/agents/loki_agent.rb]: I wonder if it's better to pass the entire expression as a parameter to this method – that way the agent isn't tied to namespaces, as such. So this line would just be:

```json
"expr": expression
```

File: app/agents/loki_agent.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                          }
+                        ]'
+    if grafana_url.end_with?('/')
+      @grafana_url + '/' + CGI.escape(logging_view_uri)
+    else
+      @grafana_url + CGI.escape(logging_view_uri)
+    end
+  end
+
+  def delete_logging_view(_namespace)
```
Comment (jits) [inline app/agents/loki_agent.rb]:
```suggestion
  def delete_logging_dashboard(name)
```

File: app/workers/resources/request_create_worker.rb [hunk 6]
```diff
@@ -45,6 +45,17 @@ class RequestCreateWorker < BaseWorker
...
           resource.url = result.url
 
+          true
+        end
+      },
+      'Resources::LoggingDashboard' => {
+        'loki' => lambda do |resource, agent, config|
+          grafana_url = config['grafana_url']
+
+          result = agent.create_logging_view resource.name, grafana_url: grafana_url
```
Comment (jits) [inline app/workers/resources/request_create_worker.rb]: Doesn't look like you need `grafana_url` here?

File: app/workers/resources/request_create_worker.rb [hunk 7]
```diff
@@ -45,6 +45,17 @@ class RequestCreateWorker < BaseWorker
...
+
+          resource.url = result.url
```
Comment (jits) [inline app/workers/resources/request_create_worker.rb]: `result` is the URL value

File: app/agents/loki_agent.rb (hunk 4 above)
Comment (danielwhatmuff) [reply app/agents/loki_agent.rb]: yes good shout

File: app/workers/resources/request_create_worker.rb [hunk 8]
```diff
@@ -45,6 +45,18 @@ class RequestCreateWorker < BaseWorker
...
 
+          true
+        end
+      },
+      'Resources::LoggingDashboard' => {
+        'loki' => lambda do |resource, agent, config|
+          grafana_url = config['grafana_url']
+          query_expression = '{"namespace":\"' + resource.name + '\"}'
+
+          result = agent.create_logging_dashboard resource.name, query_expression, grafana_url: grafana_url
```
Comment (jits) [inline app/workers/resources/request_create_worker.rb]: You still don't need the `grafana_url` here :)

**Pull request closed without merging**